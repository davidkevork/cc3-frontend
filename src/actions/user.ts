import 'whatwg-fetch';
import { action } from 'typesafe-actions';
import { API_URL } from '../constants/const';

export const UserActionTypes = {
  USER_FROM_TOKEN: 'USER_FROM_TOKEN',
  USER_FROM_TOKEN_SUCCESS: 'USER_FROM_TOKEN_SUCCESS',
  USER_FROM_TOKEN_FAILURE: 'USER_FROM_TOKEN_FAILURE',

  LOGOUT_SUCCESS: 'LOGOUT_SUCCESS',
  LOGOUT_FAILURE: 'LOGOUT_FAILURE',

  RESET_USER_TOKEN: 'RESET_USER_TOKEN',
  SET_USER_LOADING: 'SET_USER_LOADING',
  SET_USER_ERROR: 'SET_USER_ERROR',
  CLEAR_USER_ERROR: 'CLEAR_USER_ERROR',
}

/** User from token */
export const userFromToken = (
  tokenFromStorage: string,
  userId: string,
): any => action(UserActionTypes.USER_FROM_TOKEN, fetch(`${API_URL}users/${userId}`, {
  method: 'GET',
  headers: new Headers({
    Accept: 'application/json',
    'Content-Type': 'application/json',
    Authorization: `Bearer ${tokenFromStorage}`,
  }),
}).then(response => response.json()));

export const userFromTokenFailure = (
  error: Error,
) => action(UserActionTypes.USER_FROM_TOKEN_FAILURE, error);

export const userFromTokenSuccess = (
  user: object,
) => action(UserActionTypes.USER_FROM_TOKEN_SUCCESS, user);

/** Logout */
export const logoutSuccess = () => action(UserActionTypes.LOGOUT_SUCCESS, null);

export const logoutFailure = (
  error: Error,
) => action(UserActionTypes.LOGOUT_FAILURE, error);

/** Common actions */
export const resetUserToken = (
) => action(UserActionTypes.RESET_USER_TOKEN);

export const setUserLoading = (
) => action(UserActionTypes.SET_USER_LOADING);

export const setUserError = (
  error: Error,
) => action(UserActionTypes.SET_USER_ERROR, error);

export const clearUserError = (
) => action(UserActionTypes.CLEAR_USER_ERROR);
