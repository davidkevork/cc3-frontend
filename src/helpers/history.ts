import { createBrowserHistory } from 'history';

export const history = createBrowserHistory();

export function redirect(to: string) {
  window.location.href = to;
}
