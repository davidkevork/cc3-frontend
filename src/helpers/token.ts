import jwt from 'jsonwebtoken';

export interface ItokenObject {
  token: string;
  userId: string;
  exp: number;
  jti: string;
  error: string | undefined;
}

export function getToken(token: string = ''): ItokenObject {
  const data: ItokenObject = {
    token: '',
    userId: '',
    exp: 0,
    jti: '',
    error: undefined,
  };

  let decode: { [key: string]: any } = {};

  if (token === '') {
    try {
      const storageToken: string | null = localStorage.getItem('jwtToken');
      data.token = storageToken === null ? '' : storageToken;
    } catch (e) {
      data.error = 'Cant get token';
      return data;
    }
  } else {
    data.token = token;
  }

  if (data.token === '') {
    data.error = 'No token found';
    return data;
  }

  try {
    const storageDecode: string | { [key: string]: any } | null = jwt.decode(data.token);
    if (storageDecode !== null && typeof storageDecode !== 'string') {
      decode = storageDecode;
    } else {
      data.error = 'Invalid token. error 1';
      return data;
    }
  } catch (err) {
    data.error = 'Invalid token. error 1';
    return data;
  }

  if (decode.permission !== undefined && decode.permission.userId !== undefined) {
    data.userId = decode.permission.userId;
  } else {
    data.error = 'Invalid token. error 2';
    return data;
  }

  if (decode.exp !== undefined) {
    data.exp = decode.exp;
  } else {
    data.error = 'Invalid token. error 3';
    return data;
  }

  if (decode.jti !== undefined) {
    data.jti = decode.jti;
  } else {
    data.error = 'Invalid token. error 4';
    return data;
  }

  return data;
}

export function removeToken(): boolean {
  try {
    localStorage.removeItem('jwtToken');
    return true;
  } catch (err) {
    return false;
  }
}

export function setToken(token: string = ''): boolean {
  try {
    localStorage.setItem('jwtToken', token);
    return true;
  } catch (err) {
    return false;
  }
}
