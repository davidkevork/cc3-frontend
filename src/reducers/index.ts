import { combineReducers } from 'redux';
import { IUserReducerState, UserReducer } from './reducer_user';

export interface RootState {
  user: IUserReducerState;
}

const rootReducer = combineReducers({
  user: UserReducer,
});

export default rootReducer;
