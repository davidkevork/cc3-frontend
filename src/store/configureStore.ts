import { Store, createStore, applyMiddleware } from 'redux';
import promise from 'redux-promise';
import rootReducer, { RootState } from '../reducers';

const enhancer = applyMiddleware(
  promise,
);

export const configureStore = (initialState?: any): Store<RootState> => {
  return createStore(rootReducer, initialState, enhancer);
};
