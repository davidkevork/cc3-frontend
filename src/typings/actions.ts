export interface valueAction<T> {
  type: string;
  payload: T;
}