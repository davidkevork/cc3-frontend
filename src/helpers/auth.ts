import { Dispatch } from 'redux';
import { toast } from 'react-toastify';
import * as user from '../actions/user';
import { redirect } from './history';
import { getToken, removeToken, ItokenObject } from './token';
import { valueAction } from '../typings/actions';
import {
  UserObject,
} from '../typings/reducer';

type authType = 'private' | 'public';

export function publicAuth(
  dispatch: Dispatch,
  props: any,
  callback: (success: boolean) => void,
) {
  const data = getToken();
  if (data.error === undefined) {
    if (props.jwt === data.token && props.userId === data.userId) {
      callback(true);
      redirect('/home');
    } else {
      userAuth(dispatch, data, 'public', callback);
    }
  } else {
    removeToken();
    clearLocalStorage();
    dispatch(user.resetUserToken());
    callback(false);
  }
}

export function privateAuth(
  dispatch: Dispatch,
  props: any,
  callback: (success: boolean) => void,
) {
  const data = getToken();
  if (data.error !== undefined) {
    removeToken();
    clearLocalStorage();
    dispatch(user.resetUserToken());
    callback(false);
    redirect('/login');
  } else if (
    props.jwt === data.token &&
    props.userId === data.userId
  ) {
    callback(true);
  } else {
    userAuth(dispatch, data, 'private', callback);
  }
}

export function userAuth(
  dispatch: Dispatch,
  data: ItokenObject,
  type: authType,
  callback: (success: boolean) => void,
) {
  dispatch(user.setUserLoading());
  dispatch(user.userFromToken(data.token, data.userId))
    .then((response: valueAction<UserObject>) => {
      if (
        response.payload.items !== undefined &&
        response.payload.status === 200
      ) {
        const parseData = response.payload.items[0];
        dispatch(
          user.userFromTokenSuccess({
            jwt: data.token,
            userId: data.userId,
            email: parseData.email,
          }),
        );
        callback(true);
        if (type === 'public') {
          redirect('/home');
        }
      } else if (response.payload.status === 401) {
        clearLocalStorage();
        dispatch(user.logoutSuccess());
        callback(false);
        if (type === 'private') {
          redirect('/login');
        }
      } else if (response.payload.error !== undefined) {
        removeToken();
        clearLocalStorage();
        dispatch(
          user.userFromTokenFailure(response.payload.error.message),
        );
        callback(false);
        if (type === 'private') {
          redirect('/login');
        }
      } else {
        callback(false);
      }
    })
    .catch((e: Error) => {
      callback(false);
    });
}

export function logout(dispatch: Dispatch) {
  toast.info('Logging out');
  removeToken();
  clearLocalStorage();
  dispatch(user.resetUserToken());
  redirect('/login');
}

export function clearLocalStorage() {
  localStorage.removeItem('jwtToken');
}
