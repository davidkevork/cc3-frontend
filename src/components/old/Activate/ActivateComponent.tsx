import React from 'react';
import { hot } from 'react-hot-loader';
import intl from 'react-intl-universal';
import IntlMessageFormat from 'intl-messageformat';
import { redirect } from '../../helpers/history';
import { UserBusinessActionTypes } from '../../actions/userBusiness';
import { IUserBusinessReducerState } from '../../reducers/reducer_user_business';
import IndexNav from '../../IndexNav/IndexNav';
import { enUS } from '../../locales/en-US';
import './Activate.scss';

interface IActivateComponentState {
  token: string | undefined;
  __html: string;
}

interface IActivateComponentProps {
  activate: (token: string) => void;
  match: {
    params: {
      token: string | undefined;
    },
  };
  userBusiness: IUserBusinessReducerState;
}

class ActivateComponent extends React.Component<IActivateComponentProps, IActivateComponentState> {
  constructor(props: IActivateComponentProps) {
    super(props);
    this.state = {
      token: this.props.match.params.token,
      __html: intl.get('ACTIVATE.PLEASE_WAIT').d(enUS.ACTIVATE.PLEASE_WAIT),
    };
    if (this.state.token !== undefined) {
      this.props.activate(this.state.token);
    }
  }
  public componentWillReceiveProps(nextProps: IActivateComponentProps) {
    if (nextProps.userBusiness.status === UserBusinessActionTypes.ACTIVATE_FAILURE
      && typeof nextProps.userBusiness.error === 'string') {
      this.setState({ __html: nextProps.userBusiness.error });
    } else if (nextProps.userBusiness.status === UserBusinessActionTypes.ACTIVATE_SUCCESS) {
      nextProps.userBusiness.role === 'user' ? redirect('/cards') : redirect('/locations');
    }
  }
  public render() {
    // tslint:disable-next-line:max-line-length
    const type = process.env.ROLE === 'user' ? intl.get('ACTIVATE.CARD').d(enUS.ACTIVATE.CARD) : intl.get('ACTIVATE.BUSINESS').d(enUS.ACTIVATE.BUSINESS);
    return (
      <React.Fragment>
        <IndexNav />
        <div className="page-activate">
          {this.state.token === undefined ?
            (
              <div className="activate-box">
                <i className="icon-sent-mail" />
                <h3 className="text-center">
                  {intl.get('ACTIVATE.VERIFY_EMAIL').d(enUS.ACTIVATE.VERIFY_EMAIL)}
                </h3>
                <p className="text-center">
                  {intl.get('ACTIVATE.ALMOST_THERE', { type }).d(
                    new IntlMessageFormat(enUS.ACTIVATE.ALMOST_THERE, 'en-US').format({ type }),
                  )}
                </p>
              </div>
            )
          :
            (
              <div className="activate-box">
                <h3 className="text-center" dangerouslySetInnerHTML={this.state} />
              </div>
            )
          }
        </div>
      </React.Fragment>
    );
  }
}

export default hot(module)(ActivateComponent);
