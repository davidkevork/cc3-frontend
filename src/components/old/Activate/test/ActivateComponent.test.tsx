import * as React from 'react';
import * as enzyme from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import promise from 'redux-promise';
import ActivateComponent from '../ActivateComponent';
import { UserBusinessReducer } from '../../../reducers/reducer_user_business';
import { UserBusinessActionTypes } from '../../../actions/userBusiness';
import * as history from '../../../helpers/history';

const mockStore = configureStore([promise]);

describe('Activate Component', () => {
  it('should render correctly', () => {
    const store = mockStore({ userBusiness: UserBusinessReducer });
    let ActivateComponentRendered = enzyme.shallow(
      <ActivateComponent
        activate={() => {}}
        match={{
          params: {
            token: undefined,
          }
        }}
        userBusiness={store.userBusiness}
      />
    );
    expect(toJson(ActivateComponentRendered)).toMatchSnapshot();
  });
  it('should activate user/business', () => {
    const activateMock = jest.fn();
    const store = mockStore({ userBusiness: UserBusinessReducer });
    let userBusiness = store.userBusiness;
    let ActivateComponentUserBusiness = enzyme.shallow(
      <ActivateComponent
        activate={activateMock}
        match={{
          params: {
            token: 'random-id',
          }
        }}
        userBusiness={userBusiness}
      />
    );
    expect(toJson(ActivateComponentUserBusiness)).toMatchSnapshot();
    expect(activateMock).toBeCalledWith('random-id');
  });
  it('should receive props', async () => {
    const store = mockStore({ userBusiness: UserBusinessReducer });
    let ActivateComponentReceiveProps = enzyme.shallow(
      <ActivateComponent
        activate={() => {}}
        match={{
          params: {
            token: undefined,
          }
        }}
        userBusiness={store.userBusiness}
      />
    );

    await ActivateComponentReceiveProps.instance().componentWillReceiveProps({
      userBusiness: { status: UserBusinessActionTypes.ACTIVATE_FAILURE, error: 'random error' },
    }, ActivateComponentReceiveProps.state());
    expect(ActivateComponentReceiveProps.state('__html')).toStrictEqual('random error');

    let redirectSpy = jest.spyOn(history, 'redirect');
    
    await ActivateComponentReceiveProps.instance().componentWillReceiveProps({
      userBusiness: { status: UserBusinessActionTypes.ACTIVATE_SUCCESS },
    }, ActivateComponentReceiveProps.state());
    expect(redirectSpy).toHaveBeenCalledTimes(1);
    
    redirectSpy.mockReset();
    redirectSpy.mockRestore();
  });
});