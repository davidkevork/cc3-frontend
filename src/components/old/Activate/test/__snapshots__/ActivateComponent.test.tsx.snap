// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Activate Component should activate user/business 1`] = `
<Fragment>
  <Connect(IndexNav) />
  <div
    className="page-activate"
  >
    <div
      className="activate-box"
    >
      <h3
        className="text-center"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Please wait while we activate your account",
            "token": "random-id",
          }
        }
      />
    </div>
  </div>
</Fragment>
`;

exports[`Activate Component should render correctly 1`] = `
<Fragment>
  <Connect(IndexNav) />
  <div
    className="page-activate"
  >
    <div
      className="activate-box"
    >
      <i
        className="icon-sent-mail"
      />
      <h3
        className="text-center"
      >
        Verify your email address
      </h3>
      <p
        className="text-center"
      >
        You're almost there. Please finalise your business account creation by verifying your email with the activation email we just sent.
      </p>
    </div>
  </div>
</Fragment>
`;
